### PR TITLE
GitHub

### PR DESCRIPTION
**Pull request by RafalDrozdz**

### CODE REVIEW

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import UserPopout from "../../UserPopup";
+
+const Contributor = ({ contributor }) => {
+    const { user } = useContributor(contributor.author.url);
+    console.log(user);
+
+    const [popup, setPopup] = React.useState(false);
+    return (
+        <div className="contributor">
+            <div className="contributor__link" onClick={() => setPopup(true)}>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: Please use React hooks. ```
const handleClick = React.useCallback((){
 setPopup(true)
},[])
```

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+                <div className="contributor__avatar">
+                    <img src={user.avatar_url} alt={user.name} />
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: it's not a proper alt.

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            <div className="contributor__commits">
+                {contributor.total > 1
+                    ? `${contributor.total} commits`
+                    : `${contributor.total} commit`}
+            </div>
+            <UserPopout popup={popup} setPopup={setPopup} user={user} />
+        </div>
+    );
+};
+
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: Create at least two separate components.

File: src/components/BaseComponents/Contributors/Contributor/useContributor.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+export function useContributor(contributor_url) {
+    const [user, setUser] = React.useState("");
+    React.useEffect(() => {
+        const getContributorDetails = async () => {
+            const data = await getGithubUser(contributor_url);
+            setUser(data);
+        };
+        getContributorDetails();
+    }, []);
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/useContributor.js:12]: I like it!

File: src/components/BaseComponents/Contributors/useContributors.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import { getContributors } from "../../../api";
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/useContributors.js]: as we discussed - change paths

File: src/components/BaseComponents/UserPopup/UserPopup.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                        <a
+                            href={user.html_url}
+                            target="_blank"
+                            rel="noreferrer"
+                        >
+                            Go to Github
+                        </a>
+                    </div>
+                </>
+            ) : null}
```
Comment (SawickiKamil) [inline src/components/BaseComponents/UserPopup/UserPopup.js]: `{popup && (...) }` in this case you don't need null. It looks bad. Divide that component and `onClick={() => setPopup(false)}` dont use logic inside the JSX, create conts with callback outside.

File: src/components/GithubRepos/GithubRepos.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import * as React from "react";
+import { useGithubRepos } from "./useGithubRepos";
+import RepoCompopnent from "./RepoCompopnent";
+
+const GithubRepos = () => {
+    const { repos } = useGithubRepos();
+    const reposItems = [...repos].map((item) => (
+        <RepoCompopnent key={item.id} repo={item} />
```
Comment (SawickiKamil) [inline src/components/GithubRepos/GithubRepos.js]: I like it!

File: src/components/GithubRepos/useGithubRepos.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import { getGithubRepos } from "../../api/";
+
+export function useGithubRepos() {
+    const [repos, setRepos] = React.useState("");
+    React.useEffect(() => {
```
Comment (SawickiKamil) [inline src/components/GithubRepos/useGithubRepos.js]: React.useState(""); is this a string? Just asking :)

File: src/components/GithubUser/GithubUser.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    const { user } = useGithubUser("RafalDrozdz");
+    const { id } = user;
+
+    return (
+        <>
+            <div>{id}</div>
+        </>
+    );
+};
+
```
Comment (SawickiKamil) [inline src/components/GithubUser/GithubUser.js]: ?

File: src/components/GithubUser/useGithubUser.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import * as React from "react";
+import { getGithubUser } from "../../api/";
+
+export function useGithubUser(username) {
+    const [user, setUser] = React.useState("");
+    React.useEffect(() => {
```
Comment (SawickiKamil) [inline src/components/GithubUser/useGithubUser.js]: const [user, setUser] = React.useState(""); is user a string or an object?

**Review by SawickiKamil — CHANGES_REQUESTED** [review]

**Review by SawickiKamil — CHANGES_REQUESTED** [review]
What is going to appear on my screen while we have no committers and the repo is empty?

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 11]
```diff
@@ -1,25 +1,54 @@
...
+    contributor.weeks.forEach((element) => {
+        added += element.a;
+        deleted += element.d;
+    });
+    React.useEffect(() => {
+        function handleClickOutside(event) {
+            if (popupRef.current && !popupRef.current.contains(event.target)) {
+                setPopup(false);
+            }
+        }
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: You are inside of the functional component feel free to use React.useCallback instead.

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 12]
```diff
@@ -1,25 +1,54 @@
...
+        document.addEventListener("mousedown", handleClickOutside);
+        return () => {
+            document.removeEventListener("mousedown", handleClickOutside);
+        };
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: Nice, this is exactly the way how the component life cycle works!

File: src/components/BaseComponents/Contributors/Contributor/Contributor.js [hunk 13]
```diff
@@ -1,25 +1,54 @@
...
+            <div className="contributor__container__commits">
+                <div className="contributor__commits">
+                    {contributor.total > 1
+                        ? `${contributor.total} commits`
+                        : `${contributor.total} commit`}
+                </div>
+                <div className="contributor__lines">
+                    <div className="contributor__added">{added}++</div>
+                    <div className="contributor__deleted">{deleted}--</div>
+                </div>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/Contributor.js]: Would be nice to keep in mind it is really difficult to test this component. Create two separate ones

File: src/components/BaseComponents/UserPopup/UserPopup.js [hunk 14]
```diff
@@ -2,31 +2,27 @@ import * as React from "react";
...
-                </>
+                    <div className="popup__image">
+                        <img src={user.avatar_url} alt={user.login} />
+                    </div>
+                    <div className="popup__login">{user.login}</div>
+                    <div className="popup__bio">{user.bio}</div>
+                    <div className="popup__location">{user.location}</div>
+                    <a href={user.html_url} target="_blank" rel="noreferrer">
+                        Go to Github
+                    </a>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/UserPopup/UserPopup.js]: can we create a separate component? UserDetails?

File: src/styles/components/contributor.scss [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        color: #aaa;
+        font-size: fontClamp(0.9, 1.8);
+    }
+}
+
+@media (max-width: 1024px) {
+    .contributor {
+        width: 96%;
+    }
+}
```
Comment (SawickiKamil) [inline src/styles/components/contributor.scss:65]: No clue. I believe it is a proper way of creating styles :P

**Review by SawickiKamil — CHANGES_REQUESTED** [review]
What is going to appear on my screen while we have no committers and the repo is empty?

File: src/components/BaseComponents/UserPopup/UserPopup.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    return (
+        <>
+            {popup ? (
+                <div className="popup">
+                    <div
+                        className="popup__close"
+                        onClick={() => setPopup(false)}
+                    >
+                        x
+                    </div>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/UserPopup/UserPopup.js]: Would be super cool to implement a complex solution. We should close the modal by clicking outside the box. The easiest way - a third-party library, but you can learn something by implementing it on your own :)

File: src/components/BaseComponents/UserPopup/UserPopup.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+                    <div className="popup__container__right">
+                        <a
+                            href={user.html_url}
+                            target="_blank"
+                            rel="noreferrer"
+                            className="popup__link"
+                        >
+                            Go to Github
+                        </a>
+                    </div>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/UserPopup/UserPopup.js]: Could we create two separate components here? Usually, the link should be reusable in many places. Try to think globally. It's not a big issue in that small project but take it into account that mostly you will need to work on an enormous project with thousand of components and tests. If you'll decide to chunk the project, the advantage of this solution will be invaluable.

**Review by SawickiKamil — CHANGES_REQUESTED** [review]

**Comment by SawickiKamil** [comment]
https://medium.com/@janelle.wg/atomic-design-pattern-how-to-structure-your-react-application-2bb4d9ca5f97

File: .gitignore [hunk 18]
```diff
@@ -18,6 +18,8 @@ public
 .env.development.local
 .env.test.local
 .env.production.local
+.env
+.env.development
```
Comment (SawickiKamil) [inline .gitignore]: Do you need both?

File: gatsby-config.js [hunk 19]
```diff
@@ -15,5 +20,17 @@ module.exports = {
...
+        {
+            resolve: "gatsby-plugin-root-import",
+            options: {
+                api: path.join(__dirname, "src/api"),
+                styles: path.join(__dirname, "src/styles"),
+                components: path.join(__dirname, "src/components"),
+                db: path.join(__dirname, "src/db"),
+                assets: path.join(__dirname, "src/assets"),
+                hooks: path.join(__dirname, "src/hooks"),
+                constans: path.join(__dirname, "src/constans"),
```
Comment (SawickiKamil) [inline gatsby-config.js:32]: Can we simplify?

File: src/components/BaseComponents/Avatar/Avatar.js [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import * as React from "react";
+
+const Avatar = ({ src, alt, className = "popup__image", children }) => {
+    return (
+        <div className={className}>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Avatar/Avatar.js]: check out classnames

Comment (SawickiKamil) [reply src/components/BaseComponents/Avatar/Avatar.js]: className={classnames{["popup__image", className]}}

File: src/components/BaseComponents/Contributors/Contributor/ContributorDetails/ContributorLink/ContributorLink.js [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+
+const ContributorLink = ({ user, handleClick }) => {
+    return (
+        <div className="contributor__link" onClick={handleClick}>
+            <div className="contributor__avatar">
+                <img
+                    src={user.avatar_url}
+                    alt={`${user.login} - user avatar`}
+                />
+            </div>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Contributors/Contributor/ContributorDetails/ContributorLink/ContributorLink.js]: separate component?

File: src/components/BaseComponents/Link/Link.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
+import * as React from "react";
+import "styles/components/popup.scss";
+
+const Link = ({
+    href,
+    target = "_blank",
+    rel = "noreferrer",
```
Comment (SawickiKamil) [inline src/components/BaseComponents/Link/Link.js:7]: nice one

File: src/components/BaseComponents/RepoLanguages/RepoLanguage/RepoLanguage.js [hunk 23]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+
+const RepoLanguage = ({ item, languageNumber }) => {
+    return (
+        <div className="languages__language">
+            <div className="languages__languageText">{item.language}</div>
+            <div className="languages__procents">
+                {`${((item.number / languageNumber) * 100).toFixed(1)}%`}
```
Comment (SawickiKamil) [inline src/components/BaseComponents/RepoLanguages/RepoLanguage/RepoLanguage.js]: magic numbers? const something = tem.number / languageNumber) * 100).toFixed(1)}%

File: src/components/BaseComponents/RepoLanguages/RepoLanguages.js [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+import RepoLanguage from "./RepoLanguage";
+import RepoLanguagesBar from "./RepoLanguagesBar";
+import { useRepoLanguages } from "./useRepoLanguages";
+
+import "styles/components/repoLanguages.scss";
+
+const RepoLanguages = ({ repoLanguagesUrl }) => {
+    const { languages } = useRepoLanguages(repoLanguagesUrl);
+
+    const languagesArray = Object.keys(languages).map((key) => ({
```
Comment (SawickiKamil) [inline src/components/BaseComponents/RepoLanguages/RepoLanguages.js]: it's too generic

File: src/components/BaseComponents/RepoLanguages/RepoLanguagesBar/RepoLanguagesBar.js [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+const RepoLanguagesBar = ({ languagesArray, languageNumber }) => {
+    return (
+        <div className="languages__bar">
+            {languagesArray.map((item) => (
+                <div
+                    key={item.language}
+                    className={`languages__bar__${item.language}`}
+                    style={{
+                        width: `${(item.number / languageNumber) * 100}%`,
```
Comment (SawickiKamil) [inline src/components/BaseComponents/RepoLanguages/RepoLanguagesBar/RepoLanguagesBar.js]: Magic numbers

File: src/components/BaseComponents/UserPopup/UserPopupRight/UserPopupRight.js [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+
+const UserPopupRight = ({ url }) => {
+    return (
+        <>
+            <div className="popup__container__right">
+                <Link href={url} className="popup__link">
+                    Go to Github
+                </Link>
+            </div>
+        </>
```
Comment (SawickiKamil) [inline src/components/BaseComponents/UserPopup/UserPopupRight/UserPopupRight.js]: we don't need RF

File: src/hooks/useOpenClose.js [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+export function useOpenClose() {
+    const [open, setOpen] = React.useState(false);
+
+    const handleClickOpen = React.useCallback(() => {
+        setOpen(true);
+    }, []);
+
+    const handleClickClose = React.useCallback(() => {
+        setOpen(false);
+    }, []);
```
Comment (SawickiKamil) [inline src/hooks/useOpenClose.js]: [setOpen]

File: src/pages/github.js [hunk 28]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+
+import GithubRepos from "components/GithubRepos";
+
+import "styles/pages/github.scss";
+
+const Github = () => (
+    <>
+        <GithubRepos />
+    </>
```
Comment (SawickiKamil) [inline src/pages/github.js]: RF?

**Review by SawickiKamil — APPROVED** [review]
LGTM

File: gatsby-config.js (hunk 19 above)
Comment (RafalDrozdz) [reply gatsby-config.js:32]: I was searching but I did't find anything

**Review by SawickiKamil — APPROVED** [review]
LGTM! Looks lit!

**Pull request merged**